src/Main.hs

```diff
@@ -83,6 +94,31 @@ noEdges = AllEdges { normalEdges    = IMap.empty
                    , sourceEdges    = IMap.empty
                    }
 
+combs :: Int -> [([Int], [Int])] -> [[([Int], [Int])]]
+combs 0 _ = [[]]
+combs _ [] = []
+combs k (x:xs) = ((x :) <$> combs (k - 1) xs) ++ combs k xs
+
+toposort :: [(Int, [Int])] -> [Int]
+toposort xs
+  | (not . null) cycleDetect =
+    error $ "Dependency cycle detected for nodes " ++ show cycleDetect
+  | otherwise = foldl makePrecede [] dB
+  where
+    dB :: [([Int], [Int])]
+    dB = ((\(x, y) -> (x, y \\ x)) . (return *** id)) <$> xs
```

Comment:
Could you simplify this to avoid using the funny combinators?  A list comprehension would be sufficient.